src/btree/bt_ret.c

```diff
@@ -247,10 +241,8 @@ __wt_value_return_buf(WT_CURSOR_BTREE *cbt, WT_REF *ref, WT_ITEM *buf, WT_TIME_W
         if (tw != NULL)
             WT_TIME_WINDOW_COPY(tw, &unpack.tw);
         ret = __wt_page_cell_data_ref(session, page, &unpack, buf);
-#ifdef HAVE_DIAGNOSTIC
         if (ret == 0 && unpack.cell != NULL && is_ovfl_rm != NULL)
             *is_ovfl_rm = __wt_cell_type_raw(unpack.cell) == WT_CELL_VALUE_OVFL_RM;
```

Comment:
Further, if the cell type can change while traversing to this bit of code (due to a reconciliation), is it possible that copying the content out (not seeing a WT_CELL_VALUE_OVFL_RM here, but having it change to one at a different time) will leave a window where invalid or freed memory is referenced?

Reply:
This is not possible because of the ovfl_lock
```
    /*
     * WT_CELL_VALUE_OVFL_RM cells: if reconciliation deletes an overflow value, the on-page cell
     * type is reset to WT_CELL_VALUE_OVFL_RM. Any values required by an existing reader will be
     * copied into the HS file, which means this value should never be read. It's possible to race
     * with checkpoints doing that work, lock before testing the removed flag.
     */
    __wt_readlock(session, &S2BT(session)->ovfl_lock);
    if (__wt_cell_type_raw(unpack->cell) == WT_CELL_VALUE_OVFL_RM) {
        ret = __wt_buf_setstr(session, store, "WT_CELL_VALUE_OVFL_RM");
        *decoded = true;
    } else
        ret = __ovfl_read(session, unpack->data, unpack->size, store);
    __wt_readunlock(session, &S2BT(session)->ovfl_lock);
```

Reply:
From the code, when we read an overflow value, we just have a pointer to the location of the overflow page in memory. If the overflow page is discarded from memory, then we would be pointing to freed memory. I'm not sure when we discard overflow value from memory though. It seems it is discarded along with the page that references it?

Reply:
```
    /*
     * WT_CELL_VALUE_OVFL_RM cells: if reconciliation deletes an overflow value, the on-page cell
     * type is reset to WT_CELL_VALUE_OVFL_RM. Any values required by an existing reader will be
     * copied into the HS file, which means this value should never be read. It's possible to race
     * with checkpoints doing that work, lock before testing the removed flag.
     */
    __wt_readlock(session, &S2BT(session)->ovfl_lock);
    if (__wt_cell_type_raw(unpack->cell) == WT_CELL_VALUE_OVFL_RM) {
        ret = __wt_buf_setstr(session, store, "WT_CELL_VALUE_OVFL_RM");
        *decoded = true;
    } else
        ret = __ovfl_read(session, unpack->data, unpack->size, store);
    __wt_readunlock(session, &S2BT(session)->ovfl_lock);
```
The overflow lock ensures that before we finish reading the overflow block into memory, the state of the cell cannot change. When reading the overflow block, we always copy it in full to a temporary buffer in memory. We are not freeing the in memory buffer when we set the cell to removed. The buffer is discarded later. So it is not possible for us to have a window that will read freed memory.